Extension multitenancy documentation (#289)
AxonIQ/reference-guide: extensions/multitenancy.md

> #Multitenancy Extension

Reviewer: Since this is a new page, we need to add it to the index of the Reference Guide too.
You can do this by adjusting the `SUMMARY.md` file.

> #Multitenancy Extension

Reviewer: This page should be part of the 4.6 documentation, since that's when we introduced it.
For that to happen, you need to point the pull request to the `4.6` branch, instead of `master`.

Note that adjusting this just may require you to recreate this pull request.

Axon Framework Multitenancy Extension provides your application ability to serve multiple tenants (event-stores) at once.
Multi-tenancy is important in cloud computing and this extension will provide ability to connect to tenant dynamically, physical separate tenants data, scale tenants independently...

### Requirements

> Following requirements needs to be meet for extension to work:

Author: true, I have added that

> ** Axon Cloud **
>
> Axon Cloud works only with static tenant configuration.

### Configuration

Minimal configuration is needed to get extension up and running.
Choose to use **either** static or dynamic tenant configuration.

#### Static tenants configuration

If you have predefined list of contexts that your application should connect to, set following property:
`axon.axonserver.contexts=tenant-context-1,tenant-context-2,tenant-context-3`

#### Dynamic tenants configuration

If you plan to create tenants in runtime, you can define a predicate which will tell application to which contexts to connect to once they appear in runtime:

```java
    @Bean
public TenantConnectPredicate tenantFilterPredicate() {
        return context -> context.tenantId().startsWith("tenant-");
        }
```

### Route message to specific tenant

#### Using meta-data

By default, to route message to specific tenant you need to tag initial message that enters your system with metadata.
This is done with meta-data helper, and to route message to specific tenant you should set tenant name to metadata with key `TenantConfiguration.TENANT_CORRELATION_KEY`.

```java

> message.andMetaData(Collections.singletonMap(TENANT_CORRELATION_KEY, "tenant-context-1")

Author: We did configure an automatic meta data resolver (as default),  but the user still needs to set the initial meta data to the message, this is something that we can't do as we don't know how the user wants to distinguish tenants, or from where to get this info

```

Metadata needs to be added only to initial message that enters your system. Any message that is produced by consequence of initial message will have this metadata copied automatically using to `CorrelationProvider`.

#### Custom tenant resolver

If you wish to define custom tenant resolver set following property:

`axon.multi-tenancy.use-metadata-helper=false`

Then define custom tenant resolver bean. For example following bean can use message payload to route message to specific tenant:

> ```java

Author: unfortunately yes but not for long, one of the first issues that Il pick up in this extension is non-spring configured

If you wish to implement multi-tenancy for a different type of databases (e.g. NoSQL) make sure that your projection database supports multi-tenancy.
While in transaction you may find out which tenant owns transaction by calling:` TenantWrappedTransactionManager.getCurrentTenant()`.

> **Pre initialize schema**
>
> Schema migration tools like Liquibase or Flyway usually won't be able to initialise schemas for dynamically created data sources.
> Any datasource that will use needs to have pre-initialized schema.
 
#### Resetting projections

Resetting projections works a bit different because there are multiple instances of "same" event processor group (one per each tenant).

Reset specific tenant event processor group:

```java
    TrackingEventProcessor trackingEventProcessor = configuration.eventProcessingConfiguration()
                                                                     .eventProcessor("com.demo.query-ep@tenant-context-1",
                                                                                     TrackingEventProcessor.class)
                                                                     .get();
```

> Convention for naming event processor is: `{even processor name}@{tenant name}`

Reviewer: ```suggestion
Note that the convention for naming tenant-specific event processor is `{even processor name}@{tenant name}`.
```

> Access all tenant event processors by retrieving `MultiTenantEventProcessor` only.
> `MultiTenantEventProcessor` acts as a proxy Event Processor that references all tenant event processors.

Reviewer: ```suggestion
If you need to access all tenant event processors in one go, you can retrieve the `MultiTenantEventProcessor` for a specific processing name.
The `MultiTenantEventProcessor` acts as a proxy event processor referencing all tenant-specific event processors.
```

### Supported multi-tenant components

> Currently, supported multi-tenants components are as follows:

Reviewer: ```suggestion
Currently, the following infrastructure components support multi-tenancy:
```

- <span style="color:green">MultiTenantCommandBus</span>
- <span style="color:green">MultiTenantEventProcessor</span>
- <span style="color:green">MultiTenantEventStore</span>
- <span style="color:green">MultiTenantQueryBus</span>
- <span style="color:green">MultiTenantQueryUpdateEmitter</span>
- <span style="color:green">MultiTenantEventProcessorControlService</span>
- <span style="color:green">MultiTenantDataSourceManager</span>

> Not yet supported multi-tenants components are:

Reviewer: ```suggestion
The following components are not yet supported:
```

- <span style="color:red">MultitenantDeadlineManager</span>
- <span style="color:red">MultitenantEventScheduler</span>

> ### Disable extension

Reviewer: ```suggestion
### Disabling this Extension
```

> By default, extension is automatically enabled if found on class path.
> If you wish to disable extension without removing extension use following property
> 
> `axon.multi-tenancy.enabled=false`

Reviewer: ```suggestion
By default, this extension is enabled if found on class path when utilizing Spring Boot.
If you wish to disable the extension without removing the dependency, you can set the following property to `false`:

`axon.multi-tenancy.enabled=false`
```
